File: src/hes/sudoku/Cell.java
package hes.sudoku;

import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Set;

public class Cell {

    private Integer number = 0;
    private boolean known = false;
    private Set<Integer> candidates = new LinkedHashSet<>(9);

    private String name;
    private Set<Group> groups = new LinkedHashSet<>();

    public Cell(String name) {
        this.name = name;
        for (int number = 1; number <= 9; number++) {
            candidates.add(number);
        }
    }

    public boolean isKnown() {
        return known;
    }

    public Integer getNumber() {
        return number;
    }

    public Set<Integer> getCandidates() {
        return candidates;
    }

    public void setNumber(Integer number) {
        this.number = number;
        known = true;
        candidates = Collections.EMPTY_SET;
    }

    public void eliminate(Integer number) {
        candidates.remove(number);
    }

    public Set<Group> getGroups() {
        return groups;
    }

    public String text() {
        if (known) {
            return String.valueOf(number) + " ";
        }

        return ". ";
    }

    public int nrOfCandidates() {
        return candidates.size();
    }

    public Integer firstCandidate() {
        return candidates.iterator().next();
    }

    public String allCandidatesAsString() {
        if (known) {
            return String.format("        .");
        }

        StringBuilder builder = new StringBuilder(9);
        candidates.forEach(builder::append);

        return String.format("%9s", builder);
    }

    @Override
    public String toString() {
        return name;
    }

    public boolean hasCandidate(Integer number) {
        return candidates.contains(number);
    }
}

File: src/main/java/hes/sudoku/kt/Cell.kt
package hes.sudoku.kt

class Cell {

    val row: Int
    val column: Int
    var number = 0
        private set

    private val _candidates = mutableSetOf<Int>()
    private val _groups = mutableSetOf<Group>()

    // Derived properties

    val known: Boolean
        get() = number != 0

    val candidates: Set<Int> = _candidates
    val groups: Set<Group> = _groups

    val name: String
        get() = "Row ${row + 1} column ${column + 1}"

    val text: String
        get() = if (known) "$number " else ". "

    val candidatesAsString: String
        get() {
            if (known) {
                return "        ."
            }

            val builder = StringBuilder(9)
            _candidates.forEach { builder.append(it) }

            return String.format("%9s", builder)
        }

    override fun toString(): String {
        return name
    }

    //
    // Init
    //

    constructor(row: Int, column: Int) {
        this.row = row
        this.column = column

        for (number in 1..9) _candidates.add(number)
    }

    constructor(original: Cell) {
        number = original.number
        _candidates.addAll(original._candidates)
        row = original.row
        column = original.column
    }

    fun add(group: Group) {
        _groups.add(group)
    }

    // For unit tests
    internal fun replaceCandidates(candidates: Set<Int>): Cell {
        this._candidates.clear()
        this._candidates.addAll(candidates)

        return this
    }

    //
    // Methods
    //

    fun setNumber(number: Int) {
        this.number = number
        _candidates.clear()
    }

    fun eliminate(number: Int?) {
        _candidates.remove(number)
        check(!(!known && _candidates.size == 0)) { "No more candidates for $this after eliminating %$number" }
    }
}